src/utils.rs

```diff
@@ -112,6 +120,136 @@ pub async fn wait_tx_in_finalized_block(
+		.with_context(|| format!("write {}", path.display()))?;
+	writer.flush().with_context(|| format!("flush {}", path.display()))?;
+
+	println!("Wrote JSON data to {}", path.display());
+	Ok(())
+}
+
+/// Read data from a JSON file
+pub async fn read_data_from_json_file<T>(file_path: &str) -> anyhow::Result<T>
+where
+	T: DeserializeOwned,
+{
+	let path = Path::new(file_path);
+	println!("Reading data from file: {:?}", path);
+
+	let mut file = File::open(path)
+		.with_context(|| format!("file not found at {}", path.display()))?;
+	let mut content = String::new();
+	file
+		.read_to_string(&mut content)
+		.with_context(|| format!("failed to read {}", path.display()))?;
+
+	serde_json::from_str(&content)
+		.with_context(|| format!("failed to deserialize {}", path.display()))
+}
+
+/// Chain properties from system_properties RPC call
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct ChainProperties {
+	token_symbol: Option<String>,
+	token_decimals: Option<u8>,
+}
+
+/// Get the SS58 prefix from the chain
+pub async fn get_ss58_prefix(client: &Client) -> Result<u16, Error> {
+	match crate::dynamic::pallet_api::system::constants::SS58_PREFIX.fetch(client.chain_api()) {
+		Ok(ss58_prefix) => Ok(ss58_prefix),
+		Err(e) => {
+			log::warn!(target: LOG_TARGET, "Failed to fetch SS58 prefix: {}", e);
+			log::warn!(target: LOG_TARGET, "Using default SS58 prefix: 0");
+			Ok(0)
+		},
+	}
+}
+
+/// Get chain properties (token decimals and symbol) from system_properties RPC
+pub async fn get_chain_properties(client: &Client) -> Result<(u8, String), Error> {
+	// Create a new RPC client for the call
+	let rpc_client = subxt::backend::rpc::RpcClient::from_url(client.uri())
+		.await
+		.map_err(|e| Error::Other(format!("Failed to create RPC client: {}", e)))?;
+
+	// Call system_properties RPC method
+	let response: ChainProperties = rpc_client
+		.request("system_properties", rpc_params![])
+		.await
+		.map_err(|e| Error::Other(format!("Failed to call system_properties RPC: {}", e)))?;
+
+	// Extract token decimals (first value from array)
+	let decimals = response.token_decimals.unwrap_or(10); // Default to 10 for most Substrate chains
+
+	// Extract token symbol (first value from array)
+	let symbol = response.token_symbol.unwrap_or("UNIT".to_string()); // Default symbol
+
+	log::info!(
+		target: LOG_TARGET,
+		"Fetched chain properties: token_symbol={}, token_decimals={}",
+		symbol,
+		decimals
+	);
+
+	Ok((decimals, symbol))
+}
+
+/// Encode an AccountId to SS58 string with chain-specific prefix
+pub fn encode_account_id(account: &AccountId, ss58_prefix: u16) -> String {
+	// AccountId is already AccountId32, just encode it with the custom prefix
+	// Clone to avoid move
+	account
+		.clone()
+		.to_ss58check_with_version(Ss58AddressFormat::custom(ss58_prefix))
+}
+
+/// Convert Plancks to tokens (divide by 10^decimals) and format with token symbol
+pub fn planck_to_token(planck: u128, decimals: u8, symbol: &str) -> String {
```

Comment:
Have you seen this crate? it can likely simplify this code a bit for you: https://github.com/paritytech/ss58-registry